### PR TITLE
fix(browse): exponential backoff when OpenAI throttles API calls

### PR DESCRIPTION
**Pull request by dbillor**
adding retry/backoff ability by pulling function out

### CODE REVIEW

File: scripts/browse.py [hunk 1]
```diff
@@ -80,36 +82,72 @@ def create_message(chunk, question):
...
 
-    text_length = len(text)
-    print(f"Text length: {text_length} characters")
+@retry(wait=wait_exponential(multiplier=1, min=4, max=30), stop=stop_after_attempt(5))
+def call_create_chat_completion(messages):
+    return create_chat_completion(
+        model=cfg.fast_llm_model,
+        messages=messages,
+        max_tokens=300,
+
```
Comment (brotherko) [inline scripts/browse.py]: Missing the closing ")"

File: scripts/browse.py [hunk 2]
```diff
@@ -80,36 +82,72 @@ def create_message(chunk, question):
...
 
+def summarize_text(text, is_website=True):
+    if text == "":
```
Comment (sberney) [inline scripts/browse.py]: Why the change from `not text` to `== ""`?

File: scripts/browse.py [hunk 3]
```diff
@@ -80,36 +82,72 @@ def create_message(chunk, question):
...
     }
 
-def summarize_text(text, question):
-    if not text:
-        return "Error: No text to summarize"
 
-    text_length = len(text)
-    print(f"Text length: {text_length} characters")
+@retry(wait=wait_exponential(multiplier=1, min=4, max=30), stop=stop_after_attempt(5))
+def call_create_chat_completion(messages):
```
Comment (sberney) [inline scripts/browse.py]: I would suggest moving this into the create_chat_completion file and naming it something like create_chat_completion_with_retry.

File: scripts/browse.py [hunk 4]
```diff
@@ -80,36 +82,72 @@ def create_message(chunk, question):
...
-            max_tokens=300,
-        )
-        summaries.append(summary)
-
-    print(f"Summarized {len(chunks)} chunks.")
+        print("Summarizing chunk " + str(i + 1) + " / " + str(len(chunks)))
+        if is_website:
+            messages = [
+                {
+                    "role": "user",
```
Comment (sberney) [inline scripts/browse.py]: Your spacing and indentation is messed up in this section, and inconsistent with the rest of the project.

File: scripts/browse.py [hunk 5]
```diff
@@ -80,36 +82,72 @@ def create_message(chunk, question):
...
+                    "content": "Please summarize the following website text, do not describe the general website, but instead concisely extract the specific information this subpage contains.: " +
+                    chunk},
+            ]
+        else:
+            messages = [
```
Comment (sberney) [inline scripts/browse.py]: This stuff doesn't follow the design pattern in this file. Currently this kind of material exists in methods, for example `create_message`. I would move this logic up into the class as a method(s) and try to keep the implementation in this part of the file as similar to the original in form and function as possible, except for calling a different method.

**Review by sberney — CHANGES_REQUESTED** [review]

File: scripts/browse.py (hunk 1 above)
Comment (dbillor) [reply scripts/browse.py]: ty oops >.<  (this is not OK)

File: scripts/browse.py (hunk 2 above)
Comment (dbillor) [reply scripts/browse.py]: ok

File: scripts/browse.py (hunk 3 above)
Comment (dbillor) [reply scripts/browse.py]: done

File: scripts/browse.py (hunk 4 above)
Comment (dbillor) [reply scripts/browse.py]: updated.

File: scripts/browse.py (hunk 5 above)
Comment (dbillor) [reply scripts/browse.py]: updated.

File: scripts/browse.py [hunk 6]
```diff
@@ -106,10 +109,69 @@ def summarize_text(text, question):
...
+
+
+
+
+
+
+
+
+
+
```
Comment (sberney) [inline scripts/browse.py]: A lot of extra blank lines here

**Review by sberney — COMMENTED** [review]
lgtm. Can we have a CLI flag to turn this on and off?

File: scripts/browse.py [hunk 7]
```diff
@@ -80,6 +82,7 @@ def create_message(chunk, question):
         "content": f"\"\"\"{chunk}\"\"\" Using the above text, please answer the following question: \"{question}\" -- if the question cannot be answered using the text, please summarize the text."
     }
 
+
```
Comment (nponeccop) [inline scripts/browse.py]: This blank line is not necessary

File: scripts/browse.py [hunk 8]
```diff
@@ -1,7 +1,9 @@
 import requests
 from bs4 import BeautifulSoup
 from config import Config
-from llm_utils import create_chat_completion
+from llm_utils import create_chat_completion, create_chat_completion_with_retry
+from tenacity import retry, wait_exponential, stop_after_attempt
+
```
Comment (nponeccop) [inline scripts/browse.py]: Remove this blank line

**Review by nponeccop — CHANGES_REQUESTED** [review]
LGTM but needs some whitespace changes

**Review by nponeccop — CHANGES_REQUESTED** [review]
Remove the extra whitespace

File: scripts/browse.py (hunk 8 above)
Comment (dbillor) [reply scripts/browse.py]: removed

File: scripts/browse.py (hunk 7 above)
Comment (dbillor) [reply scripts/browse.py]: removed

File: scripts/browse.py (hunk 6 above)
Comment (dbillor) [reply scripts/browse.py]: removed

File: scripts/browse.py [hunk 9]
```diff
@@ -93,8 +94,8 @@ def summarize_text(text, question):
     for i, chunk in enumerate(chunks):
         print(f"Summarizing chunk {i + 1} / {len(chunks)}")
         messages = [create_message(chunk, question)]
-
-        summary = create_chat_completion(
+
```
Comment (nponeccop) [inline scripts/browse.py]: There is trailing whitespace here

**Review by nponeccop — CHANGES_REQUESTED** [review]
Just 1 place now

**Review by nponeccop — DISMISSED** [review]

**Comment by nponeccop** [comment]
@dbillor There are conflicts now

**Comment by cryptidv** [comment]
😨

**Comment by Pwuts** [comment]
@dbillor please fix the commit history so we can merge

**Comment by dbillor** [comment]
> @dbillor please fix the commit history so we can merge

done

**Comment by AndresCdo** [comment]
@dbillor fix formatting using isort.

**Comment by dbillor** [comment]
> @dbillor fix formatting using isort.

updated

**Comment by Pwuts** [comment]
Sorry to disappoint now, but it seems this was already fixed in #1537:
https://github.com/Significant-Gravitas/Auto-GPT/blob/14d3ecaae75a80f03cd118b5767e1c6affb4e3cc/autogpt/llm_utils.py#L96

Also, this PR is currently broken as it applies the edit to `scripts/browse.py` which doesn't exist anymore.. That decorator is nice though. :)

Closing as duplicate.
